update from the latest upstream changes

r-xue opened the pull request.

r-xue commented on auto_selfcal.py:
```diff
@@ -656,6 +667,40 @@
+         ##
+         os.system('rm -rf '+sani_target+'_'+band+'_'+solint+'_'+str(iteration)+'_post*')
+         tclean_wrapper(vislist,sani_target+'_'+band+'_'+solint+'_'+str(iteration)+'_post',
+                  band_properties,band,telescope=telescope,nsigma=selfcal_library[target][band]['nsigma'][iteration], scales=[0],
```
@jjtobin I start using this PR to review the changes happening after the point I forked.

First question: in these changes from @psheehan, we removed the start model keyword in the prototype. Was this result of the discussion with Luke back in the f2f meeting? The PL version lacks this currently, I plan to get it in soon. This will certainly make some difference in the outcome.

psheehan replied:
Yes, this was the result of discussions with Luke and Ryan back in the f2f meeting. We decided that it would be best to redo tclean using the same threshold rather than using the startmodel because there were times that the start model no longer did a good job of representing the data. To get some speed back, we then copy the _post images to be the starting point for the next solint (and continue cleaning from there) so that we don't have to have a whole additional tclean from scratch.

r-xue replied:
Perfect. Thanks for the clarification! I have sent John a note that the PL version now has all recent changes from the  [jjtobin:main](https://github.com/jjtobin/auto_selfcal). So this is in!

psheehan replied:
Excellent!

jjtobin committed:
do not multiply VLA FOV by 2x for low frequenciess when doing mosaics

jjtobin committed:
fix typo and print per-field mosaic fluxes when removing field from selfcal

jjtobin committed:
disallow mosaic field removal in inf_EB and scan_inf and enable separate flux threshold for VLA

jjtobin committed:
if all mosaic fields removed in solint, fail solint

jjtobin committed:
add if statement to catch VLA spws that are not in cont.dat

jjtobin committed:
add if statement to catch VLA spws that are not in cont.dat

jjtobin committed:
adjust writing of applycal and uvcontsub calls to select correct datacolumn depending if all sources failed selfcal or if only one out of several

jjtobin committed:
Merge pull request #94 from jjtobin/fix_uvcontsub_for_unsuccessful_selfcal

adjust writing of applycal and uvcontsub calls to select correct data column

jjtobin committed:
enable rectangular mosaics and better centering of images with non rectangular mosaics

jjtobin committed:
only do brute-force mosaic detection and field aggregation for VLA

jjtobin committed:
fix indent

jjtobin committed:
attempted fix on fetch_targets

jjtobin committed:
accidental exclusion of some crucial ACA metadata

psheehan committed:
If a VLA dataset has a cont.dat, make sure to match only based on properties.

jjtobin committed:
update tests to exclude new and incompatible dictionary entries

jjtobin committed:
variable name mis-specified, triggered when cont.dat does not have all spws

jjtobin committed:
variable name mis-specified, triggered when cont.dat does not have all spws

jjtobin committed:
Merge branch 'make_pip_installable_VLA_mosaics' of https://github.com/jjtobin/auto_selfcal into make_pip_installable_VLA_mosaics

psheehan committed:
Merge remote-tracking branch 'john/main' into better_spw_mapping

austronomer committed:
Added try-except to handle IndexErrors

austronomer committed:
Adjusted print statements

jjtobin committed:
fix mis-specified variable to reference proper spectral window when matching two MSes

jjtobin committed:
update test function call again to omit certain key comparisons in the github tests

jjtobin committed:
fix more issues that omit processing mosaics with ACA data

psheehan committed:
Upload selfcal_library reference and weblogs as artifacts

psheehan committed:
Also upload casa log to artifacts

psheehan committed:
Make sure telescope is now passed to get_fitspw_dict

psheehan committed:
Update test datasets for the GitHub Actions tests

psheehan committed:
Properly update GitHub Actions test datasets

psheehan committed:
Fix deprecated numpy functionality causing mosaics to fail

psheehan committed:
Merge branch 'jjtobin:main' into fix_removed_numpy_functionality

jjtobin committed:
fixes for all subfields being removed in a selfcal solint

jjtobin committed:
Merge branch 'make_pip_installable_VLA_mosaics' of https://github.com/jjtobin/auto_selfcal into make_pip_installable_VLA_mosaics

jjtobin committed:
Merge pull request #96 from austronomer/main

Added try-except to split_calibrated_final.py

jjtobin committed:
Merge pull request #91 from psheehan/better_spw_mapping

Better handling of SPW mapping

psheehan committed:
Merge branch 'jjtobin:main' into fix_removed_numpy_functionality

psheehan committed:
Merge remote-tracking branch 'origin/main' into make_pip_installable+better_spw_mapping

psheehan committed:
Merge branch 'make_pip_installable+better_spw_mapping' into make_pip_installable

psheehan committed:
Merge remote-tracking branch 'downstream/fix_removed_numpy_functionality' into make_pip_installable

psheehan committed:
Make sure to get the vislist before removing the corrected column, as it now needs to be pulled from selfcal_library

psheehan committed:
Merge remote-tracking branch 'origin/main' into make_pip_installable+EB_alignment

psheehan committed:
Merge remote-tracking branch 'upstream/make_pip_installable' into make_pip_installable+EB_alignment

jjtobin committed:
Merge pull request #98 from psheehan/fix_removed_numpy_functionality

Fix removed numpy functionality

psheehan committed:
Add new benchmark tests

psheehan committed:
Update README to reflect pip-installable version.

psheehan committed:
Add citation guidance to README and docs, and add quickstart guide to docs

psheehan committed:
Fix docs link in README

psheehan committed:
Update quickstart Python install instructions.

psheehan committed:
Note that CASA must be available for python version for python environment install.

jjtobin committed:
unblock mosaic field removal inf_EB and scan_inf and make flux threshold for field removal a parameter and different defaults for VLA/ALMA

jjtobin committed:
Merge remote-tracking branch 'origin/make_pip_installable' into make_pip_installable_VLA_mosaics

jjtobin committed:
fix conflicts with upstream merge

jjtobin committed:
exclude testing flux_threshold key

jjtobin committed:
Merge pull request #99 from jjtobin/make_pip_installable

Restructure the code into installable Python package format

psheehan committed:
Run tests and PyPI uploads on main now that the pip-installable version has made it to main

psheehan committed:
Name in pyproject.toml should match what is on PyPI

jjtobin committed:
expose overlap_tol as a free parameter

psheehan committed:
Change pip install invocation and update reference

psheehan committed:
Fix author ordering

jjtobin committed:
update test to exclude overlap_tol

psheehan committed:
Add K-band-mini-mosaic VLA mosaic selfcal test

jjtobin committed:
added ability for inf_EB to split itself into shorter solints

jjtobin committed:
also print out the actual interval, in addition to the solint names

jjtobin committed:
fix wrong variable name for solint intervals

psheehan committed:
Add reference values for K-band-mini-mosaic test

jjtobin committed:
fix missing argument to get_SNR_self_individual

jjtobin committed:
remove unnecessary argument for applycal_wrapper

psheehan committed:
Enable better compatibility for pip-installing into monolithic CASA, including getting working with mpicasa properly, and adding a function to set up the proper symlinks.

psheehan committed:
Create an msmdWrapper class that can cache values resturned from msmd for future use without having to re-open the data.

jjtobin committed:
fix bug in metadata association with field ids

jjtobin committed:
Merge branch 'make_pip_installable_VLA_mosaics' of https://github.com/jjtobin/auto_selfcal into make_pip_installable_VLA_mosaics

psheehan committed:
Update K-band-mini-mosaic reference values

psheehan committed:
Update docs to reflect improvement in pip-installing into monolithic CASA.

jjtobin committed:
change and to or to not ALMA and ACA

jjtobin committed:
Merge branch 'make_pip_installable_VLA_mosaics' of https://github.com/jjtobin/auto_selfcal into make_pip_installable_VLA_mosaics

jjtobin committed:
update plot_image to handle rectangular images properly and remove *_old functions from selfcal_helpers.py

psheehan committed:
Update the tests to use monolithic CASA with pip-installed auto_selfcal properly.

psheehan committed:
Need to pip install --upgrade pytest to get the monolithic CASA symlink working properly.

psheehan committed:
Add pytest-cov as a dependency when asking for the [test] install.

jjtobin committed:
fix some various errors reported by users in orig_ms_helpers.py and split_final_calibrated.py

psheehan committed:
Further updates to the docs to provide more details about installing/running pip-installed into monolithic CASA.

jjtobin committed:
Merge pull request #102 from jjtobin/fix_orig_ms_issues

fix some various errors reported by users

jjtobin committed:
revert to square images and old get_phasecenter for more isolated testing

jjtobin committed:
backout unintended nfrms adjustment

jjtobin committed:
changes to capture field metadata per MS for mosaics and single targets

jjtobin committed:
fixes to get per vis information for mosaics and field names

jjtobin committed:
fix overwriting of dictionary for single fields in bands_for_targets

samdf96 committed:
Explicitly import pipeline functions to prevent NameError for findcont implementation

Although launching casa in pipeline mode does import the respective pipeline functions into the casa environment, it seems those are not passed into the python namespace. Even the pipeline.initcli() does not import these into the namespace.

Added import statements outside of the if block to resolve.

samdf96 committed:
Clean up comments in auto_selfcal.py

Removed comments explaining pipeline import behavior.

jjtobin committed:
get ALMA field ids from msmd.fieldsforscans instead of msmd.fieldsforname which can contain phantom field ids

jjtobin committed:
adopt different field specification for splitting ALMA vs. VLA data to help with mosaic testing

jjtobin committed:
revert get_phasecenter to find field IDs the old way for ALMA to reproduce benchmark

jjtobin committed:
fix telescope name typo in get_phasecenter if statement

jjtobin committed:
fix missing field argument needed for get_phasecenter

jjtobin committed:
Merge pull request #104 from samdf96/samdf96-findcont-patch

findcont patch - adding explicit import statements for pipeline functions

jjtobin committed:
integrated bands_for_targets into selfcal_library to limit the amount of passing this dict

jjtobin committed:
update printing of bands_for_targets for debug

jjtobin committed:
Update run_E2E_test.yml

jjtobin committed:
dummy push to try and trigger tests

jjtobin committed:
missing bracket

jjtobin committed:
remove print of undefined variable

jjtobin committed:
remove bands_for_target references in tclean_wrapper in auto_selfcal.py

jjtobin committed:
remove bands_for_target references in run_selfcal in auto_selfcal.py

jjtobin committed:
remove lingering bands_for_targets references

jjtobin committed:
exclude bands_for_targets from test comparison

jjtobin committed:
Merge pull request #105 from jjtobin/VLA_mosaics_less_bands_for_targets

integrated bands_for_targets into selfcal_library

jjtobin committed:
Merge branch 'main' into make_pip_installable_VLA_mosaics

jjtobin committed:
Merge pull request #95 from jjtobin/make_pip_installable_VLA_mosaics

Add VLA mosaic capability

jjtobin committed:
switch get_image_parameters_new to get_image_parameters, new get_phasecenter method

jjtobin committed:
Merge pull request #103 from jjtobin/cache_msmd_values

Speed up calculating the spw map by caching msmd function call return values

jjtobin committed:
updated image size heuristics

jjtobin committed:
add buffer pixels

jjtobin committed:
ensure image dimensions are even numbers

jjtobin committed:
declare check_even for whole function, not just mosaics

jjtobin committed:
update get_phasecenter and get_image_parameters to use all phasecenters previously recorded

jjtobin committed:
update phasecenter and imagesize calculations to use all unique pointings from all MSes

psheehan committed:
PIPE-2937: Use in-memory CASA operations for estimate_SNR and estimate_near_field_SNR

This should cut down on the disk space usage as well as the temporary file creation.

psheehan committed:
Remove type hints for estimate_*SNR functions, as python 3.8 testing doesn't like them

psheehan committed:
Update tests to use new reference values based on the image size/phasecenter changes.

psheehan committed:
Merge remote-tracking branch 'origin/rectangular_mosaics_and_phasecenter_update' into PIPE-2937_in_memory_SNR

psheehan committed:
Add a casa_tools module with an ImageReader to mimic what is done in pipeline and get the new estimate_*SNR functions working.

psheehan committed:
Update test references again after properly updating them. Included a note about the proper way to do this for future reference.

psheehan committed:
Merge branch 'rectangular_mosaics_and_phasecenter_update' into PIPE-2937_in_memory_SNR

jjtobin committed:
initial capability of setting automask parameters

jjtobin committed:
make sure to multiply user-specified noise thresholds by nfrms/rms

jjtobin committed:
Merge pull request #106 from jjtobin/rectangular_mosaics_and_phasecenter_update

switch get_image_parameters_new to get_image_parameters, new get_phasecenter method

jjtobin committed:
removed unused get_image_parameter_old function

jjtobin committed:
Merge branch 'main' into inf_EB_shorter

psheehan committed:
Properly ask for ImageReader from casa_tools rather than casatools in one place

jjtobin committed:
fix variable typos

jjtobin committed:
Merge remote-tracking branch 'origin/main' into automask_parameters

jjtobin committed:
exclude new keys from tests

jjtobin committed:
Merge pull request #108 from jjtobin/automask_parameters

Automask parameters

jjtobin committed:
Merge pull request #100 from jjtobin/inf_EB_shorter

Shorter inf_EB solution intervals possible

jjtobin committed:
initial somewhat working VLBA support

jjtobin committed:
remove _delay from solints for solint_SNR estimates

psheehan committed:
Merge remote-tracking branch 'upstream/main' into make_pip_installable+EB_alignment

psheehan committed:
Add option to align EBs to the command line script.

psheehan committed:
Skip alignment of mosaics, as that is more complicated.

jjtobin committed:
working demonstration for VLBA data, adds amplitude selfcal trials for all the same solints as phase-only

psheehan committed:
Some bugfixes to get the code working post merging recent updates from main.

jjtobin committed:
various fixes for VLBA non-calibrator sources and avoid combinespw for non-VLBA delay solutions

jjtobin committed:
fix telescope tracking within selfcal library

jjtobin committed:
fix a mis-specification of selfcal_plan as selfcal_library

psheehan committed:
Add required dependencies for EB alignment.

jjtobin committed:
add separate function to compare delaycal tables and add ability to do true applymode=calonly

jjtobin committed:
add separate function to compare delaycal tables and add ability to do true applymode=calonly

jjtobin committed:
fixes for delay cals, remove some extreneous telescope parameter passing, add options for delay cal and short amp solints

jjtobin committed:
small fixes for VLBA fallback during delay solutions to account for fact that we only have per_bb as a fall back for per_spw delay solutions; combinespw is implicitly the same as per_bb

jjtobin committed:
backout some unnecessary changes to select_best_gaincal_mode

jjtobin committed:
change back to assuming gaintype=T to see if tests will pass

jjtobin committed:
fixes to syntax errors causing some runs to crash

jjtobin committed:
fix bug that was making all solints go to combinespw

jjtobin committed:
change messaging following select best gaincal/delay mode

jjtobin committed:
update weblog to show 4 digits after decimal for beam

jjtobin committed:
select best gaincal status outside if statement and can call undefined variable

jjtobin committed:
fix bad print statement

jjtobin committed:
exclude telescope dictionary key

psheehan committed:
Add option to choose optimizer backend for alignment so that not all three need to be run. Because of this, emcee and dynesty can be optional again. All three options now make a plot with the same name, for future weblog purposes.

psheehan committed:
Add a weblog page for the alignment.

psheehan committed:
Add documentation about EB alignment.

psheehan committed:
Update docs to include information about EB alignment.

psheehan committed:
Turn on EB alignment for GitHub Actions tests.

psheehan committed:
Add alignment_helpers weblog page, which was missed before.

psheehan committed:
Update link for Band8-7m-2 online dataset to version with one EB shifted.

psheehan committed:
Update the Band-7m-2.tar.gz link properly this time.

psheehan committed:
E2E test workflow needs to install EB alignment dependencies

psheehan committed:
Explicitly only do EB alignment if there are multiple EBs in a dataset

psheehan committed:
Get the logic for when to EB align (if asked for) correct

psheehan committed:
Get the likelihood plot shown in the alignment weblog, and make sure single EB datasets don't run through align_measurement_sets (and therefore end up with offsets and try to make the weblog).

psheehan committed:
Update the Band8-7m-2 test dataset to include the correct reference values and weblog, with EB alignment added.

psheehan committed:
Delete MS files after benchmark runs are done to save space - they can take up quite a lot of it.

psheehan committed:
Remove some unused code

jjtobin committed:
avoid spw combination ambiguity for single spw datasets

psheehan committed:
Add exoALMA acknowledgement to README and docs.

jjtobin committed:
fixes for n_spws >1

jjtobin committed:
remove unused code for runtime modification of gaincal parameters and using gaintype GSPLINE

jjtobin committed:
fix if statements in gaincal_wrapper that may have inadvertantly be setting the solution modes improperly

jjtobin committed:
fix final_mode entry for a delay that does per_bb fallback

jjtobin committed:
add some code comments

jjtobin committed:
fix for delay solints that only have one mode

jjtobin committed:
allow for amplitude solutions to be pre-applied

jjtobin committed:
Merge pull request #107 from jjtobin/PIPE-2937_in_memory_SNR

PIPE-2937: Use in-memory CASA operations for estimate_SNR and estimate_near_field_SNR

jjtobin committed:
Merge pull request #111 from jjtobin/delete_benchmark_MSfiles

Delete MS files after benchmark runs are done to save space

jjtobin committed:
only do combinespw fallback if it is in the modes to attempt and only do per_bb fallback for VLBA

jjtobin committed:
decouple VLBA and do_delay_cal parameter and iscalibrator and do_delay_cal, thus do_delay_cal must now be set if it is desired

jjtobin committed:
Merge branch 'main' into vlba_improvements

jjtobin committed:
Merge pull request #112 from jjtobin/vlba_improvements

Vlba improvements

jjtobin committed:
fix new numpy incompatibility with implicit conversion of single element arrays to scalar

jjtobin committed:
rename deprecated numpy.trapz to numpy.trapezoid

psheehan committed:
Add Ryan Loomis and Sam Federman to citation, and add a .zenodo.json to exert more control over the Zenodo listing (including listing contributors who have not formally committed.

psheehan committed:
Merge branch 'main' into make_pip_installable+EB_alignment

psheehan committed:
Remove telescope from EB-alignment tclean_wrapper calls now that it is passed through the selfcal_library.

psheehan committed:
Fix incorrect author listing.

psheehan committed:
Add a test that includes a usermask and usermodel.

psheehan committed:
Add reference values for 2019.1.00691.S_SB test

jjtobin committed:
Merge pull request #110 from psheehan/make_pip_installable+EB_alignment

Add optional EB alignment prior to self-calibration

jjtobin committed:
Merge pull request #101 from jjtobin/pip_install_monolithic

Improve functionality of auto_selfcal pip-installed into monolithic CASA

psheehan committed:
Merge branch 'main' into add_usermodelmask_test

psheehan committed:
Bump version number to 2.1.0 in preparation for the next release.

psheehan committed:
Fix a few errors that cause multi-EB mosaics with spws that change across the EBs to crash.

psheehan committed:
Remove unneeded print statement that was added for debugging.

jjtobin committed:
fixes to get auto_selfcal to run cleanly with allow_cocal=True

jjtobin committed:
changes to get cocal to look at original targets ms for possible calibrators

jjtobin committed:
fix missing argument for a replace() call

jjtobin committed:
fix another missing second argument on a replace()

jjtobin committed:
missing additions to selfcal_plan solint_interval

jjtobin committed:
more progress toward getting cocal to run

jjtobin committed:
fix variable name used twice and another fail statement invalid for co-cal

jjtobin committed:
fix more double use of mode variable

jjtobin committed:
pass preapply_targets_own_inf_EB to gaincal_wrapper

jjtobin committed:
fix the target and vis file specification for cocal gaincal calls

jjtobin committed:
add a concat step to add co-calibrators to the MS of the source needing cocal

psheehan committed:
Make sure to increment iteration for all possible endings of cocal solints.

psheehan committed:
Loop over all possible gaincal modes and look up the correct calibration files for the inf_fb3 solint.

jjtobin committed:
add sanitize string calls to construction of concat vislist

jjtobin committed:
remove mask check if doing cocal

jjtobin committed:
Merge pull request #114 from jjtobin/add_usermodelmask_test

Add a test that includes a usermask and usermodel.

psheehan committed:
Update cocal to work better with setting the solint_settings for the fb solints. This includes removing the stapling together of existing gaintables for inf_fb3, and also splitting cocal preparation into prepare_cocal.

psheehan committed:
Merge branch 'fix_cocal' of https://github.com/jjtobin/auto_selfcal into fix_cocal

psheehan committed:
Preapply inf_EB for inf_fb3, not inf. Also, delete some unused, commented out code.

psheehan committed:
We need a manual override of using the same pre-apply gaintable for gaincal as is used for applycal when doing the inf_fb* modes, because in those cases the same table might not be used for both.

jjtobin committed:
add sanitize_string() to target and cal_target when doing a .replace on a filename

jjtobin committed:
Merge branch 'fix_cocal' of https://github.com/jjtobin/auto_selfcal into fix_cocal

jjtobin committed:
write selfcal_plan.pickle before starting selfcal and ensure single_spw data have a solution populated for all solints

psheehan committed:
Add dataset to test cocal online.

psheehan committed:
Merge branch 'fix_cocal' of https://github.com/jjtobin/auto_selfcal into fix_cocal

psheehan committed:
Merge branch 'main' into fix_cocal

psheehan committed:
Fix 'n_spws' => 'nspws' bug; update cocal test link

psheehan committed:
Add reference values/weblog for online cocal test

psheehan committed:
Update README badges for the main repository instead of a fork

jjtobin committed:
added VLBA tests one online, one offline to tests

jjtobin committed:
fix syntax error

jjtobin committed:
fix if statement for identifying VLBA and make wget quiet to reduce logs

jjtobin committed:
fix missing quote

jjtobin committed:
add print statment regarding what file is being downloaded

jjtobin committed:
Temporarily omit some tests and fix vlba test

jjtobin committed:
Update vlba test

jjtobin committed:
add fix from fix_cocalfor single spectral window case

jjtobin committed:
update VLBA test with reference values

jjtobin committed:
reinstate commented tests

jjtobin committed:
Merge pull request #117 from jjtobin/add_VLBA_test

added VLBA tests one online, one offline to tests

jjtobin committed:
Merge branch 'main' into fix_cocal

jjtobin committed:
re-enable cocal on github tests

psheehan committed:
Force reinstall pytest during installation into monolithic CASA in case the pytest delivered is already up-to-date

jjtobin committed:
Merge pull request #118 from jjtobin/pip_install_monolithic_CASA_with_up_to_date_pytest

Force reinstall pytest during installation into monolithic CASA in case the pytest delivered is already up-to-date

jjtobin committed:
remove extreneous parameters to bin/auto_selfcal.py

jjtobin committed:
change parallel=False back to parallel=parallel in selfcal_helpers tclean_wrapper

jjtobin committed:
Merge pull request #116 from jjtobin/fix_cocal

fixes to get auto_selfcal to run cleanly with allow_cocal=True

jjtobin committed:
Merge pull request #115 from jjtobin/fix_mosaic_disjoint_spws_error

Fix errors that arise when self-calibrating multi-EB mosaics with spws that change across the EBs
